wallet-new/src/Cardano/Wallet/Kernel/BListener.hs

```diff
@@ -77,15 +79,33 @@ switchToFork pw@PassiveWallet{..} n bs = do
     k <- Node.getSecurityParameter _walletNode
     blocksAndMeta <- mapM (prefilterBlock' pw) bs
     let (blockssByAccount, metas) = unzip blocksAndMeta
-    res <- update' _wallets $ SwitchToFork k n blockssByAccount
+
+    -- Prevent new wallet restorations from starting while we
+    -- switch to the new fork.
+    (res, restorations) <- withMVar _walletRestorationTask $ \restorations -> do
```

Comment:
It's a bit confusing why we don't need to update the `MVar` here. I think the code would be clearer if we changed `WalletRestorationInfo` to instead have an `IO` action that returns its state (stuff like current slot etc.). Then as soon as we call `reset` on any one worker, the associated `IO` action could immediately return the new state. This would then also explain why we don't need to change the value of the MVar here: we _only_ need to do that when we delete workers or create new ones.